Updates to Surrogate Visualization Module

bpaul4 opened the pull request:
## Fixes


## Summary/Motivation:
Modifications to existing surrogate visualization utilities

## Changes proposed in this PR:
- Reorganize plotting tools into main 'surrogate' and private 'non-surrogate' plotting methods
- Clean up formatting, and integrate PDF print functionality into main code for more efficient memory usage
- Minor changes in response to #599 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

andrewlee94 commented on idaes/surrogate/plotting/sm_plotter.py:
```diff
@@ -30,22 +30,30 @@
+
+
+def _scatter2D(xdata, zdata, zfit, xlabels=None, zlabels=None,
+               show=True, filename=None):
```
Comment for discussion - the private method has a `show` argument, but the public method does not. Should we include the `show` argument in the public method as well to allow for cases where the user would prefer to just save the plots to file before looking at them?

bpaul4 replied:
We could....I'll leave the default to 'True' since most users would want the plots to print to their IDE viewer.

andrewlee94 commented on idaes/surrogate/plotting/sm_plotter.py:
```diff
@@ -60,64 +68,59 @@ def scatter2D(xdata, zdata, xtest, zfit, xlabels=None, zlabels=None, dz=None,
-    pointidx = -1  # dummy index for each plot produced
-    fig, ax = [], []
+    if filename is not None:
+        if filename[-4:] != '.pdf':  # adding extension if not present
```
We might want to be careful here. This will just append `.pdf` to the filename if the last four characters are not `.pdf`. However, if the user specified an extension that was not `.pdf` we would get something like `foo.bar.pdf` as the filename which might cause some problems (depending on OS).

bpaul4 replied:
Testing some specific cases, saving to '.xps' and no extension produce files that show the plot images with no issues (the second one is strange to me), and even saving to '.txt' produces a text file successfully. I initially wanted to ensure the PdfPages() call would print to .pdf, although we could just let the user enter any name including the extension they want.

andrewlee94 replied:
In many ways, the extension itself doesn't matter - it merely tells the OS what programme to use to open the file by default. The more important issue however is how the OS would handle a case where there are two periods in the file name. I think a lot of OS's can handle that these days (not that I try), but it used to be a problem and should probably be avoided. There is also a further issue that the user might not be aware that we append `.pdf` to the file name if it is not present, and might deliberately use the name `foo.bar` and expect the resulting file to have that name.

For sanity, I would often take the most restrictive approach and instead raise and `Exception` if the filename does not end in `.pdf`. That way there is no ambiguity.

bpaul4 replied:
I did try '[name].bar.pdf' and it opened normally on my Windows machine, however I agree with you about enforcing '.pdf' in the filename.

andrewlee94 commented on idaes/surrogate/plotting/sm_plotter.py:
```diff
@@ -60,64 +68,59 @@ def scatter2D(xdata, zdata, xtest, zfit, xlabels=None, zlabels=None, dz=None,
-                                        xlabels[i] + ']')
-                ax[pointidx].set_title('2D Derivative Plot')
-                ax[pointidx].legend()
+            fig = plt.figure()
```
I am not completely familiar with the plotting tools, but this looks to me that this keeps all the plots open which will result in the memory issues @carldlaird was discussing.

bpaul4 replied:
The new form of the loop keeps the file open, but overwrites the figure/axis on each iteration rather than saving a list of figure objects as before. If we close the file after saving each figure, PdfPages() will overwrite the previous file and plot. There is another package (PyPDF) that can open files and add new pages, but it would require an additional dependency since it is not part of any already installed package.

carldlaird replied:
This is exactly what I wanted. Thanks, @bpaul4

andrewlee94 commented on idaes/surrogate/plotting/sm_plotter.py:
```diff
@@ -182,55 +195,62 @@ def parity(zdata, zfit, zlabels=None, clo=None, chi=None,
 
-def residual(xdata, e, xlabels=None, elabels=None, show=True, PDF=False,
-             filename='results_residual.pdf'):
+def _residual(xdata, error, xlabels=None, elabels=None, show=True,
```
Instead of "error", would "residual" be a better name? When I see "error", I tend to think exception first.

bpaul4 replied:
I agree, as long as it's not an issue with the argument sharing a name with the method.

andrewlee94 reviewed (commented):
A couple of comments for thoguht/discussion.

bpaul4 commented:
@andrewlee94 I've addressed your comments, and changing how the figures are saved should resolve @carldlaird's memory issue as well. The test failure is due to the lack of patch coverage for the new code, and the project coverage test passes.

carldlaird commented on idaes/surrogate/plotting/sm_plotter.py:
```diff
@@ -60,64 +69,59 @@ def scatter2D(xdata, zdata, xtest, zfit, xlabels=None, zlabels=None, dz=None,
-    fig, ax = [], []
+    if filename is not None:
+        if filename[-4:] != '.pdf':  # checking if extension is present
+            raise Exception('Filename does not end in .pdf, please amend '
```
This is fine for now, but we should be using the "os" package to get the extension instead of relying on it being the last 4 characters.

carldlaird reviewed (approved):
This looks reasonable to me.
